### PR TITLE
Add presigned url multi part upload 

### PR DESCRIPTION
**Pull request by Ram81**
### Description 

This PR uses S3 multipart upload for large file uploaded using `--large` flags. It will allow us to show upload progress using progress bars. EvalAI backend [PR](https://github.com/Cloud-CV/EvalAI/pull/3052)

### CODE REVIEW

**Comment by Ram81** [comment]
@RishabhJain2018 This PR is ready for review, I have tested the feature

File: evalai/utils/common.py [hunk 1]
```diff
@@ -133,34 +152,63 @@ def generate_random_string(length):
 def upload_file_using_presigned_url(challenge_phase_pk, file, file_type, submission_metadata={}):
     if file_type == "submission":
         url = "{}{}".format(get_host_url(), URLS.get_presigned_url_for_submission_file.value)
+        upload_complete_url = "{}{}".format(get_host_url(), URLS.complete_upload_for_submission_file.value)
```
Comment (RishabhJain2018) [inline evalai/utils/common.py:155]: `complete_upload_for_submission_file` -> `finish_upload_for_submission_file`

File: evalai/utils/common.py [hunk 2]
```diff
@@ -133,34 +152,63 @@ def generate_random_string(length):
...
     elif file_type == "annotation":
         url = "{}{}".format(get_host_url(), URLS.get_presigned_url_for_annotation_file.value)
+        upload_complete_url = "{}{}".format(get_host_url(), URLS.complete_upload_for_annotation_file.value)
```
Comment (RishabhJain2018) [inline evalai/utils/common.py:158]: `complete_upload_for_annotation_file ` -> `finish_upload_for_annotation_file `

File: evalai/utils/common.py [hunk 3]
```diff
@@ -133,34 +152,63 @@ def generate_random_string(length):
...
     url = url.format(challenge_phase_pk)
     headers = get_request_header()
 
+    # Read max 100MB chunk for multipart upload
```
Comment (RishabhJain2018) [inline evalai/utils/common.py:162]: `# Read max 100MB chunk for multipart upload` -> `# Limit to max 100 MB chunk for multipart upload`

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

**Comment by RishabhJain2018** [comment]
Also, can you please check if we can add tests using mock for S3?

**Comment by Ram81** [comment]
@RishabhJain2018 Adding tests look like more work at this point. Let's pick it up after deployment as a separate task.

**Comment by RishabhJain2018** [comment]
> Adding tests look like more work at this point. Let's pick it up after deployment as a separate task.

Let's not drop the unit tests and add them in a separate PR now itself.

> We can hardcode responses like it is being done for other APIs (but I am not sure if it is correct way to do it) though but we can't mock S3 client because we are not actually using boto3 client.

okay, hardcoding and testing should be fine as it will make sure that if any API changes, our CLI client will also change accordingly.

File: requirements.txt [hunk 4]
```diff
@@ -8,3 +8,5 @@ python-dateutil==2.7.3
 requests==2.20.0
 validators==0.12.6
 termcolor==1.1.0
+tqdm==4.49.0
```
Comment (RishabhJain2018) [inline requirements.txt:11]: Why are we adding `tqdm`, if we are not using it?

Also, I would prefer to use `tqdm` over `progressbar2` as it is more maintained and more used. Any particular reason to use `progressbar2`?

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 4 above)
Comment (Ram81) [reply requirements.txt:11]: I saw `progressbar2` gives a lot of customization which we can use later on so I thought of using it. Not sure about `tqdm`

Comment (RishabhJain2018) [reply requirements.txt:11]: I think we should use `tqdm` as it very mature and maintained. `progressbar2` suffers from issues like these mentioned here -- https://github.com/WoLpH/python-progressbar#known-issues which might create problems to users later.

**Comment by Ram81** [comment]
@RishabhJain2018 I've added the tests

File: evalai/utils/common.py [hunk 5]
```diff
@@ -133,34 +152,63 @@ def generate_random_string(length):
 def upload_file_using_presigned_url(challenge_phase_pk, file, file_type, submission_metadata={}):
     if file_type == "submission":
         url = "{}{}".format(get_host_url(), URLS.get_presigned_url_for_submission_file.value)
+        upload_complete_url = "{}{}".format(get_host_url(), URLS.finish_upload_for_submission_file.value)
```
Comment (RishabhJain2018) [inline evalai/utils/common.py:155]: My bad, as I missed these in last review --
`upload_complete_url` -> `finish_upload_url`

File: evalai/utils/common.py [hunk 6]
```diff
@@ -133,34 +152,63 @@ def generate_random_string(length):
...
     elif file_type == "annotation":
         url = "{}{}".format(get_host_url(), URLS.get_presigned_url_for_annotation_file.value)
+        upload_complete_url = "{}{}".format(get_host_url(), URLS.finish_upload_for_annotation_file.value)
```
Comment (RishabhJain2018) [inline evalai/utils/common.py:158]: `upload_complete_url` -> `finish_upload_url`

File: evalai/utils/common.py [hunk 7]
```diff
@@ -133,34 +152,63 @@ def generate_random_string(length):
...
             data = dict(data, **submission_metadata)
             response = requests.post(
                 url, headers=headers, data=data
             )
+
             if response.status_code is not HTTPStatus.CREATED:
                 response.raise_for_status()
+
+            # Update url params for multipart upload on S3
+            upload_complete_url = upload_complete_url.format(challenge_phase_pk, response.json().get("submission_pk"))
```
Comment (RishabhJain2018) [inline evalai/utils/common.py:180]: Same as above!

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

**Review by RishabhJain2018 — APPROVED** [review]
LGTM :)

**Pull request merged**